integration_tests/test_pull_properties.py

```diff
@@ -86,3 +88,140 @@ def test_pull_global_build_packages_are_excluded(self):
 
         self.assertTrue(len(state.assets['build-packages']) == 0)
         self.assertNotIn('hello=2.10-1', state.assets['build-packages'])
+
+
+class GitAssetTrackingTestCase(testscenarios.WithScenarios,
+                               integration_tests.TestCase):
+
+    scenarios = [
+        ('plain', {
+            'part_name': 'git-part',
+            'expected': {}
+        }),
+        ('branch', {
+            'part_name': 'git-part-branch',
+            'expected': {
+                'branch': 'feature',
+            }
+        }),
+        ('tag', {
+            'part_name': 'git-part-tag',
+            'expected': {
+                'tag': 'feature-tag',
+            }
+        }),
+    ]
+
+    def test_pull_git(self):
+        project_dir = 'asset-tracking'
+
+        expected_commit = _source_helpers.create_git_repo('git-source')
+        self.run_snapcraft(['pull', self.part_name], project_dir)
+
+        state_file = os.path.join(
+            self.parts_dir, self.part_name, 'state', 'pull')
+        self.assertThat(state_file, FileExists())
+        with open(state_file) as f:
+            state = yaml.load(f)
+
+        self.assertIn('source-details', state.assets)
+
+        if 'branch' in self.expected:
```

Comment:
Something where you can do without an `if/else`, your expected values are of one entry point only, so if your data structure were to be something like

```python
'expected': ('key', 'value'),

...

self.assertThat(state.assets['source-details'][self.expected[0], Equals(self.expected[1]))
```

NamedTuples might be better here. If you are going to keep that `plain` with an empy expected the most you need to do is add an `if self.expected` (even rename to `self.expected_details`).

I bet you can add `_source_helpers.create_git_repo('git-source')` into the `expected of `plain` unless we agree it needs checking for all cases.